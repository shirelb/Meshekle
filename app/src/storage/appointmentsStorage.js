import axios from "axios";
import {APP_SOCKET, SERVER_URL} from "../shared/constants";


var getUserAppointmentRequests = function (userId, userHeaders) {
    return axios.get(`${SERVER_URL}/api/appointmentRequests/user/userId/${userId}`,
        {
            headers: userHeaders,
            params: {
                status: 'requested'
            },
        })
        .then(response => {
            return response;
        })
        .catch(error => {
            console.log('get user appointment requests error ', error)
        });
};

var getUserAppointments = function (userId, userHeaders) {
    return axios.get(`${SERVER_URL}/api/appointments/user/userId/${userId}`,
        {
            headers: userHeaders,
            params: {
                status: 'set'
            },
        })
        .then(response => {
            return response;
        })
        .catch(error => {
            console.log('get user appointments error ', error)
        });
};

var getUserAppointmentById = function (userId, userHeaders, eventId) {
    return axios.get(`${SERVER_URL}/api/appointments/user/userId/${userId}`,
        {
            headers: userHeaders,
            params: {status: 'set', appointmentId: eventId}
        })
        .then(response => {
            return response;
        })
        .catch(error => {
            console.log('get user appointments by id error ', error)
        });
};

var postUserAppointmentRequest = function (userId, serviceProvider, appointmentRequest, userHeaders) {
    return axios.post(`${SERVER_URL}/api/appointmentRequests/user/request`,
        {
            userId: userId,
            serviceProviderId: serviceProvider.serviceProviderId,
            role: serviceProvider.role,
            availableTime: appointmentRequest.availableTime,
            notes: appointmentRequest.notes ? appointmentRequest.notes : '',
            subject: JSON.stringify(appointmentRequest.subject)
        },
        {
            headers: userHeaders,
        })
        .then(response => {
            APP_SOCKET.emit('userPostAppointmentRequests', {
                serviceProviderId: serviceProvider.serviceProviderId
            });

            return response;
        })
        .catch(error => {
            console.log('post user appointment request error ', error)
        });
};

var cancelAppointmentRequestById = (appointmentRequest, headers) => {
    return axios.put(`${SERVER_URL}/api/appointmentRequests/user/reject`,
        {
            userId: appointmentRequest.AppointmentDetail.clientId,
            appointmentRequestId: appointmentRequest.requestId,
        },
        {
            headers: headers,
            params: {
                status: 'rejected'
            },
        }
    )
        .then((response) => {
            APP_SOCKET.emit('userCancelAppointmentRequests', {
                serviceProviderId: appointmentRequest.AppointmentDetail.serviceProviderId,
            });

            return response
        })
        .catch((error) => {
            console.log('reject appointment request error ', error);
        });
};

var cancelAppointmentById = (appointment, headers) => {
    return axios.put(`${SERVER_URL}/api/appointments/user/cancel/userId/${appointment.AppointmentDetail.clientId}/appointmentId/${appointment.appointmentId}`,
        {},
        {
            headers: headers
        }
    )
        .then((response) => {
            APP_SOCKET.emit('userCancelAppointment', {
                serviceProviderId: appointment.AppointmentDetail.serviceProviderId,
            });

            return response
        })
        .catch((error) => {
            console.log('cancel appointment error ', error);
        });
};


export default {
    getUserAppointments,
    getUserAppointmentById,
    postUserAppointmentRequest,
    getUserAppointmentRequests,
    cancelAppointmentRequestById,
    cancelAppointmentById,
};
