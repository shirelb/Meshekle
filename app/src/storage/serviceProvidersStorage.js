import axios from "axios";
import {SERVER_URL} from "../shared/constants";

var getServiceProviders = function (userHeaders) {
    return axios.get(`${SERVER_URL}/api/serviceProviders`,
        {headers: userHeaders}
    )
        .then(response => {
            return response;
        })
        .catch(error => {
            console.log('load appointments error ', error)
        });
};


export default {getServiceProviders};
