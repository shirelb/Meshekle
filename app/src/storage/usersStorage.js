import axios from "axios";
import {SERVER_URL} from "../shared/constants";

var getUserEvents = function (userId, userHeaders) {
    return axios.get(`${SERVER_URL}/api/users/events/userId/${userId}`,
        {headers: userHeaders}
    )
        .then(response => {
            return response;
        })
        .catch(error => {
            console.log('get user events error ', error)
        });
};

var getUserById = function (userId, userHeaders) {
    return axios.get(`${SERVER_URL}/api/users/userId/${userId}`,
        {headers: userHeaders}
    )
        .then(response => {
            return response;
        })
        .catch(error => {
            console.log('get user by id error ', error)
        });
};

var getUsers = function (userHeaders) {
    return axios.get(`${SERVER_URL}/api/users`,
        {headers: userHeaders}
    )
        .then(response => {
            return response.data.filter(user => user.userId !== 1 && user.userId !== "1");
        })
        .catch(error => {
            console.log('get user by id error ', error)
        });
};

var userValidToken = function (token) {
    return axios.post(`${SERVER_URL}/api/users/validToken`,
        {
            "token": token,
        },
    );
}

var userLogin = function (userId, password) {
    return axios.post(`${SERVER_URL}/api/users/login/authenticate`,
        {
            "userId": userId,
            "password": password
        },
    );
}


export default {getUserEvents, getUserById, userValidToken, userLogin, getUsers};
