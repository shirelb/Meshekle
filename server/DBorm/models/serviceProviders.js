module.exports = (sequelize, type) => {
    return sequelize.define('ServiceProviders', {
        serviceProviderId:{
            type: type.INTEGER,
            primaryKey: true,
            autoIncrement: true,
            allowNull: false,
            required: true
        },
        role : {
            type: type.STRING,
            allowNull: false,
            required: true
        },
        userId: {
            type: type.INTEGER,
            allowNull: false,
            required: true
        },
        operationTime: {
            type: type.STRING,
            allowNull: false,
            required: true
        },
        phoneNumber: {
            type: type.STRING,
            allowNull: false,
            required: true
        },
    });
};