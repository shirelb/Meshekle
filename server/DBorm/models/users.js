module.exports = (sequelize, type) => {
    return sequelize.define('Users', {
        userId:{
            type: type.STRING,
            primaryKey: true,
            required: true
        },
        fullname: {
            type: type.STRING,
            required: true
        },
        password: {
            type: type.STRING,
            required: true
        },
        email: {
            type: type.STRING,
            required: true
        },
        mailbox: {
            type: type.INTEGER,
            required: true
        },
        cellphone: {
            type: type.STRING,
            required: true
        },
        phone: {
            type: type.STRING
        },
        bornDate: {
            type: type.DATE,
            required: true
        },
        active: {
            type: type.BOOLEAN,
            required: true,
            defaultValue: true
        },
        image: {
            type: type.BLOB,
        }
    });
};