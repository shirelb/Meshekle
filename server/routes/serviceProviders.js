var express = require('express');
var router = express.Router();
const Sequelize = require('sequelize');
const Op = Sequelize.Op;
const {Users, ServiceProviders, AppointmentRequests, Appointments, Incidents, Events} = require('../DBorm/DBorm');

/* GET users listing. */
router.get('/', function(req, res, next) {
  res.send('respond with a resource serviceProviders');
});


/* POST user . */
router.post('/users/add', function (req, res, next) {
    Users.create({
        userId: req.body.userId,
        fullname: req.body.fullname,
        password: req.body.password,
        email: req.body.email,
        mailbox: req.body.mailbox,
        cellphone: req.body.cellphone,
        phone: req.body.phone,
        bornDate: new Date(req.body.bornDate),
        active: true,
    })
        .then((newUser) => {
            console.log('New user' + newUser.username + ', with id ' + newUser.userId + ' has been created.');
            res.status(200).send({"message": "User successfully added!", newUser});
        })
        .catch(err => {
            console.log(err);
            res.status(500).send(err);
        })
});

/* POST user . */
router.post('/add', function (req, res, next) {
    ServiceProviders.create({
        serviceProviderId: req.body.serviceProviderId,
        userId: req.body.userId,
        role: req.body.role,
        operationTime: req.body.operationTime,
        phoneNumber: req.body.phoneNumber,
        appointmentWayType: req.body.appointmentWayType,
        active: true,
    })
        .then((newUser) => {
            console.log('New user' + newUser.username + ', with id ' + newUser.userId + ' has been created.');
            res.status(200).send({"message": "User successfully added!", newUser});
        })
        .catch(err => {
            console.log(err);
            res.status(500).send(err);
        })
});

module.exports = router;
