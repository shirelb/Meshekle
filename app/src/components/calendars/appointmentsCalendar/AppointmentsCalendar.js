import React, {Component} from 'react';
import {ScrollView, StyleSheet} from 'react-native';
import {Calendar, LocaleConfig} from 'react-native-calendars';
import '../localConfig';
import moment from 'moment';
import phoneStorage from "react-native-simple-store";
import appointmentsStorage from "../../../storage/appointmentsStorage";
import {APP_SOCKET} from "../../../shared/constants";
import AppointmentsDayInfo from "../../appointments/AppointmentsDayInfo";
import serviceProvidersStorage from "../../../storage/serviceProvidersStorage";

export default class AppointmentsCalendar extends Component {
    constructor(props) {
        super(props);

        this.state = {
            markedDates: {},
            selectedDate: '',
            dateModalVisible: false,
            expanded: {}
        };

        this.userHeaders = {};
        this.userId = null;
    }

    componentDidMount() {
        phoneStorage.get('userData')
            .then(userData => {
                this.userHeaders = {
                    'Authorization': 'Bearer ' + userData.token
                };
                this.userId = userData.userId;
                this.loadAppointments();
            });

        APP_SOCKET.on("getUserAppointments", this.loadAppointments.bind(this));
    }

    componentWillUnmount() {
        APP_SOCKET.off("getUserAppointmentsâ€š");
    }

    loadAppointments = () => {
        appointmentsStorage.getUserAppointments(this.userId, this.userHeaders)
            .then(response => {
                let markedDates = {};

                response.data.forEach(appointment => {
                    const date = moment(appointment.startDateAndTime).format('YYYY-MM-DD');
                    if (markedDates[date] === undefined || markedDates[date] === null) {
                        markedDates[date] = {marked: true, selected: false, appointments: []};
                    }
                    markedDates[date].appointments.push(appointment);
                });

                this.setState({
                    markedDates: markedDates
                });
            })
    }

    afterCloseModalShowSelectDay = () => {
        let updatedMarkedDates = this.state.markedDates;

        const selectedDay = updatedMarkedDates[this.state.selectedDate];
        selectedDay.selected = true;
        selectedDay.color = 'blue';
        updatedMarkedDates[selectedDay] = selectedDay;

        this.setState({markedDates: updatedMarkedDates, dateModalVisible: false})
    };

    onDaySelect = (day) => {
        let updatedMarkedDates = this.state.markedDates;

        if (this.state.selectedDate !== '') {
            let lastMarkedDate = updatedMarkedDates[this.state.selectedDate];
            lastMarkedDate.selected = false;
            updatedMarkedDates[this.state.selectedDate] = lastMarkedDate;
        }

        const selectedDay = moment(day.dateString).format('YYYY-MM-DD');
        if (updatedMarkedDates[selectedDay] === undefined)
            updatedMarkedDates[selectedDay] = {marked: true, selected: false, appointments: []};
        let newMarkedDate = updatedMarkedDates[selectedDay];
        newMarkedDate.selected = true;
        newMarkedDate.color = 'blue';
        updatedMarkedDates[selectedDay] = newMarkedDate;

        let expanded = {};
        if (updatedMarkedDates[selectedDay].appointments.length > 0)
            updatedMarkedDates[selectedDay].appointments.forEach(appointment => {
                serviceProvidersStorage.getServiceProviderUserDetails(appointment.AppointmentDetail.serviceProviderId, this.userHeaders)
                    .then(user => {
                        appointment.serviceProviderFullname = user.data.fullname;

                        expanded[appointment.appointmentId] = false;

                        this.setState({
                            selectedDate: moment(day.dateString).format('YYYY-MM-DD'),
                            markedDates: updatedMarkedDates,
                            dateModalVisible: true,
                            expanded: expanded,
                        });
                    });
            });
        else
            this.setState({
                selectedDate: moment(day.dateString).format('YYYY-MM-DD'),
                markedDates: updatedMarkedDates,
                dateModalVisible: true,
                expanded: expanded,
            });
    };


    render() {
        LocaleConfig.defaultLocale = 'il';

        return (
            <ScrollView>
                <Calendar
                    markedDates={this.state.markedDates}
                    onDayPress={this.onDaySelect}
                    onDayLongPress={this.onDaySelect}
                    style={styles.calendar}
                    // Month format in calendar title. Formatting values: http://arshaw.com/xdate/#Formatting
                    // monthFormat={'yyyy MM'}
                    // Handler which gets executed when visible month changes in calendar. Default = undefined
                    // onMonthChange={(month) => {
                    //     console.log('month changed', month)
                    // }}
                    // Replace default arrows with custom ones (direction can be 'left' or 'right')
                    // renderArrow={(direction) => (<Arrow/>)}
                    theme={{
                        backgroundColor: '#ffffff',
                        calendarBackground: '#ffffff',
                        textSectionTitleColor: '#b6c1cd',
                        selectedDayBackgroundColor: '#00adf5',
                        selectedDayTextColor: '#ffffff',
                        todayTextColor: '#00adf5',
                        dayTextColor: '#2d4150',
                        textDisabledColor: '#d9e1e8',
                        dotColor: '#00adf5',
                        selectedDotColor: '#ffffff',
                        arrowColor: 'orange',
                        monthTextColor: 'blue',
                        textDayFontFamily: 'monospace',
                        textMonthFontFamily: 'monospace',
                        textDayHeaderFontFamily: 'monospace',
                        textMonthFontWeight: 'bold',
                        textDayFontSize: 16,
                        textMonthFontSize: 16,
                        textDayHeaderFontSize: 16
                    }}
                />

                <AppointmentsDayInfo
                    dateModalVisible={this.state.dateModalVisible}
                    selectedDate={this.state.selectedDate}
                    onAppointmentRequestPress={this.props.onAppointmentRequestPress}
                    loadAppointments={this.loadAppointments}
                    markedDates={this.state.markedDates}
                    expanded={this.state.expanded}
                    afterCloseModalShowSelectDay={this.afterCloseModalShowSelectDay}
                    userHeaders={this.userHeaders}
                />
            </ScrollView>
        );
    }
}

const styles = StyleSheet.create({
    calendar: {
        borderTopWidth: 2,
        // marginTop: 5,
        paddingTop: 5,
        borderBottomWidth: 2,
        borderColor: '#eee',
        height: 350,
    },
    text: {
        textAlign: 'center',
        borderColor: '#bbb',
        padding: 10,
        backgroundColor: '#eee'
    },
});