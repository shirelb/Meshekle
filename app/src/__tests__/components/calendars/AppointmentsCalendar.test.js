import React from 'react';

import {Calendar} from 'react-native-calendars';
import {shallow} from "enzyme/build";
import AppointmentsCalendar from "../../../components/calendars/appointmentsCalendar/AppointmentsCalendar";
import usersStorage from "../../../storage/usersStorage";
import phoneStorage from "react-native-simple-store";
import users from "../../jsons/users";
import serviceProvidersStorage from "../../../storage/serviceProvidersStorage";
import serviceProviders from "../../jsons/serviceProviders";
import appointmentsStorage from "../../../storage/appointmentsStorage";
import user135985058Appointments from "../../jsons/user135985058Appointments";

jest.mock("../../../storage/serviceProvidersStorage");
jest.mock("../../../storage/usersStorage");
jest.mock("react-native-simple-store");


describe("AppointmentsCalendar should", () => {
    let wrapper = null;
    let componentInstance = null;
    const userIdTest = "135985058";
    const userTest = users.findIndex(user => user.userId === userIdTest);
    const props = {};
    const mockStore = {
        userData: {
            userId: userIdTest,
            token: "some token"
        }
    };
    const navigation = {navigate: jest.fn(), dispatch: jest.fn()};

    phoneStorage.get = jest.fn().mockImplementation((key) => Promise.resolve(mockStore[key]));
    phoneStorage.update = jest.fn().mockImplementation((key, value) => Promise.resolve(mockStore[key] = value));
    usersStorage.getUsers = jest.fn().mockImplementation(() => Promise.resolve(users));
    usersStorage.getUserByUserID = jest.fn().mockImplementation((userId) => Promise.resolve(users.filter(user => user.userId === userId)[0]));
    serviceProvidersStorage.getServiceProviders = jest.fn().mockResolvedValue(serviceProviders);
    serviceProvidersStorage.getServiceProviderUserDetails = jest.fn().mockImplementation((serviceProviderId) => Promise.resolve({data: users.filter(user => user.userId === serviceProviderId)[0]}));
    serviceProvidersStorage.getServiceProviderById = jest.fn().mockImplementation((serviceProviderId) => Promise.resolve(serviceProviders.filter(provider =>
        provider.serviceProviderId === serviceProviderId)));
    appointmentsStorage.getUserAppointments = jest.fn().mockImplementation(() => Promise.resolve({data: user135985058Appointments}));

    beforeEach(() => {
        jest.clearAllMocks();
    });

    it('match snapshot', async () => {
        wrapper = shallow(<AppointmentsCalendar {...props} />);
        expect(wrapper).toMatchSnapshot();
    });

    it("render what the user see", async () => {
        wrapper = await shallow(<AppointmentsCalendar {...props} />);
        componentInstance = wrapper.instance();

        await wrapper.update();

        expect(wrapper.find(Calendar)).toHaveLength(1);
        expect(wrapper.find('AppointmentsDayInfo')).toHaveLength(1);
    });

    it("mounted with the right data", async () => {
        expect(Object.keys(componentInstance.state.markedDates).length).toEqual(3);
        expect(componentInstance.state.selectedDate).toEqual('');
        expect(componentInstance.state.dateModalVisible).toEqual(false);
        expect(componentInstance.state.expanded).toEqual({});
    });

    it("show select day after cosed modal", async () => {
        wrapper = await shallow(<AppointmentsCalendar {...props} />);
        componentInstance = wrapper.instance();

        await wrapper.update();

        await componentInstance.setState({selectedDate: "2019-07-03"});
        await componentInstance.afterCloseModalShowSelectDay();

        expect(componentInstance.state.markedDates["2019-07-03"].selected).toEqual(true);
        expect(componentInstance.state.markedDates["2019-07-03"].color).toEqual("blue");
    });

    it("on select day show day info modal with appointment in the day", async () => {
        wrapper = await shallow(<AppointmentsCalendar {...props} />);
        componentInstance = wrapper.instance();

        await wrapper.update();

        await componentInstance.setState({selectedDate: "2019-07-03"});
        await componentInstance.onDaySelect({dateString: "2020-03-21"});

        expect(Object.keys(componentInstance.state.markedDates).length).toEqual(3);
        expect(componentInstance.state.markedDates["2020-03-21"].appointments.length).toEqual(1);
        expect(componentInstance.state.markedDates["2020-03-21"].selected).toEqual(true);
        expect(componentInstance.state.markedDates["2020-03-21"].color).toEqual("blue");
    });

    it("on select day show day info modal with out appointment in the day", async () => {
        wrapper = await shallow(<AppointmentsCalendar {...props} />);
        componentInstance = wrapper.instance();

        await wrapper.update();

        await componentInstance.setState({selectedDate: "2019-07-03"});
        await componentInstance.onDaySelect({dateString: "2021-03-21"});

        expect(Object.keys(componentInstance.state.markedDates).length).toEqual(4);
        expect(componentInstance.state.markedDates["2021-03-21"].appointments.length).toEqual(0);
        expect(componentInstance.state.markedDates["2021-03-21"].selected).toEqual(true);
        expect(componentInstance.state.markedDates["2021-03-21"].color).toEqual("blue");
        expect(componentInstance.state.selectedDate).toEqual("2021-03-21");
        expect(componentInstance.state.dateModalVisible).toEqual(true);
    });

});