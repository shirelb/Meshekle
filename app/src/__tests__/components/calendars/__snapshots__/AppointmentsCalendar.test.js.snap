// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppointmentsCalendar should match snapshot 1`] = `
<ScrollViewMock>
  <Calendar
    markedDates={Object {}}
    onDayLongPress={[Function]}
    onDayPress={[Function]}
    style={
      Object {
        "borderBottomWidth": 2,
        "borderColor": "#eee",
        "borderTopWidth": 2,
        "height": 350,
        "paddingTop": 5,
      }
    }
    theme={
      Object {
        "arrowColor": "orange",
        "backgroundColor": "#ffffff",
        "calendarBackground": "#ffffff",
        "dayTextColor": "#2d4150",
        "dotColor": "#00adf5",
        "monthTextColor": "blue",
        "selectedDayBackgroundColor": "#00adf5",
        "selectedDayTextColor": "#ffffff",
        "selectedDotColor": "#ffffff",
        "textDayFontFamily": "monospace",
        "textDayFontSize": 16,
        "textDayHeaderFontFamily": "monospace",
        "textDayHeaderFontSize": 16,
        "textDisabledColor": "#d9e1e8",
        "textMonthFontFamily": "monospace",
        "textMonthFontSize": 16,
        "textMonthFontWeight": "bold",
        "textSectionTitleColor": "#b6c1cd",
        "todayTextColor": "#00adf5",
      }
    }
  />
  <AppointmentsDayInfo
    afterCloseModalShowSelectDay={[Function]}
    dateModalVisible={false}
    expanded={Object {}}
    loadAppointments={[Function]}
    markedDates={Object {}}
    selectedDate=""
    userHeaders={Object {}}
  />
</ScrollViewMock>
`;
